posix/mount.c

```diff
@@ -0,0 +1,288 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// clang-format off
+#include <openenclave/enclave.h>
+#include <openenclave/internal/thread.h>
+// clang-format on
+
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/corelibc/string.h>
+#include <openenclave/internal/trace.h>
+#include <openenclave/corelibc/limits.h>
+#include <openenclave/corelibc/stdio.h>
+#include <openenclave/internal/posix/device.h>
+#include <openenclave/internal/posix/raise.h>
+#include <openenclave/bits/safecrt.h>
+
+#define MAX_MOUNT_TABLE_SIZE 64
+
+typedef struct _mount_point
+{
+    char* path;
+    size_t path_size;
+    oe_device_t* fs;
+    uint32_t flags;
+} mount_point_t;
+
+static mount_point_t _mount_table[MAX_MOUNT_TABLE_SIZE];
+size_t _mount_table_size = 0;
+static oe_spinlock_t _lock = OE_SPINLOCK_INITIALIZER;
+
+static bool _installed_free_mount_table = false;
+
+static void _free_mount_table(void)
+{
+    for (size_t i = 0; i < _mount_table_size; i++)
+        oe_free(_mount_table[i].path);
+}
+
+oe_device_t* oe_mount_resolve(const char* path, char suffix[OE_PATH_MAX])
+{
+    oe_device_t* ret = NULL;
+    size_t match_len = 0;
+    char realpath[OE_PATH_MAX];
+
+    if (!path || !suffix)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    /* First check whether a device id is set for this thread. */
+    {
+        uint64_t devid;
+
+        if ((devid = oe_get_thread_devid()) != OE_DEVID_NONE)
+        {
+            oe_device_t* device;
+
+            if (!(device = oe_get_device(devid, OE_DEVICE_TYPE_FILESYSTEM)))
+                OE_RAISE_ERRNO(OE_EINVAL);
+
+            /* Use this device. */
+            oe_strlcpy(suffix, path, OE_PATH_MAX);
+            ret = device;
+            goto done;
+        }
+    }
+
+    /* Find the real path (the absolute non-relative path). */
+    if (!oe_realpath(path, realpath))
+        OE_RAISE_ERRNO(oe_errno);
+
+    oe_spin_lock(&_lock);
+    {
+        /* Find the longest binding point that contains this path. */
+        for (size_t i = 0; i < _mount_table_size; i++)
+        {
+            size_t len = oe_strlen(_mount_table[i].path);
+            const char* mpath = _mount_table[i].path;
+
+            if (mpath[0] == '/' && mpath[1] == '\0')
+            {
+                if (len > match_len)
+                {
+                    if (suffix)
+                    {
+                        oe_strlcpy(suffix, realpath, OE_PATH_MAX);
```

Comment:
Is it worth it to check the return value of this?

Reply:
Good catch. I check now and added OE_RAISE_ERRNO(OE_ENAMETOOLONG) on failure.